### ISSUE TITLE
other: test normalizations

### ISSUE THREAD

**Issue by georglauterbach** [docker-mailserver/docker-mailserver#3739]
### Subject

I would like to contribute to the project

### Description

We need a follow-up of #3732. Tasks:

- [x] use `.txt` suffix for `_send_email` and `_nc_wrapper`
- [ ] in `smtp_delivery.bats`, remove the need for `existing/userX` and remove these files (ref: <https://github.com/docker-mailserver/docker-mailserver/pull/3732/files#r1438975373>)
- [x] add a new helper `_send_email_unchecked` that works like `_send_email` but has no `assert_success` directly afterward to check that all went well and add such an `assert_success` in `_send_email`
- [x] use `--auth PLAIN` instead `--auth LOGIN` ?

**Comment by georglauterbach** [comment]
Re-write `_send_email_and_get_id`

```bash
function _send_email_and_get_id() {
  [[ -v CONTAINER_NAME ]] || return 1

  local -n MAIL_ID=${1}
  shift 1

  _wait_for_empty_mail_queue_in_container
  _send_email "${@}"
  _wait_for_empty_mail_queue_in_container

  # The unique ID Postfix (and other services) use may be different in length
  # on different systems (e.g. amd64 (11) vs aarch64 (10)). Hence, we use a
  # range to safely capture it.
  MAIL_ID=$(_exec_in_container tac /var/log/mail.log              \
    | grep -E -m 1 'postfix/smtpd.*: [A-Z0-9]+: client=localhost' \
    | grep -E -o '[A-Z0-9]{9,12}' || true)

  assert_not_equal "${MAIL_ID}" ''
}
```

This avoids subshells that mess with checks.

**Comment by polarathene** [comment]
Just going to dump some notes that have since gone a bit stale but were related to the changes PRs for this issue addressed.

No expectation for anyone to read through all this, but there may be something helpful when troubleshooting the changes associated to this issue.

---

<details>
<summary>Initial swaks PR change overview (Outdated)</summary>

- Renames:
  - `test/test-files/` renamed to `test/files/`
  - `test/files/email-templates/` renamed to `test/files/emails/`
  - `test/files/auth/` and `test/files/emails/` restructured to use common file prefixes as folder names instead for grouping related files.
- Added `swaks` package to replace `nc` where possible for sending email:
  - `_send_email()` adapted to `swaks` from `nc`, `_nc_wrapper()` helper provides equivalent support where `swaks` is not compatible.
  - `test/files/emails/` removed SMTP commands, file contents only contain the `DATA` portion for providing to `swaks` via `--data`.
  - `test/files/emails/` adds an `nc_raw/` directory for files that are not `swaks` compatible, and thus must go through `_nc_wrapper()` instead of `_send_email()`.
    - `dsn/` is for `dsn.bats`, where `swaks` needs to support the equivalent of `NOTIFY=success,failure` for the `RCPT TO` SMTP command (_`--to` in `swaks`_).
    - `postscreen.bats` needs to test a misbehaving client that does not wait their turn, thus we use `_nc_wrapper()` here, since it sends all input to the destination at once.
    - `permit_docker.bats` relies on `smtp-only.txt` only as a generic mail to test toggling permission to send via network trust (`PERMIT_DOCKER`). Blocked by test file as not yet migrated to new testing format (_required to access the `_send_email` helper_).
  - `test/files/auth/` no longer has SMTP Auth (_adapted to `swaks` with `--quit-after AUTH`_). Remaining files rely on `_nc_wrapper()` for IMAP / POP3 auth related testing.

**TODO:**
- `nc_raw/smtp-ehlo.txt`: Not necessary, `swaks --quit-after FIRST-HELO`

</details>

---

## Follow-up PRs

- Mail intended to be **_submitted_ from a client _to DMS_** (_most over port 465 with auth_): `--ehlo mail.example.test` + `--from user@example.test` + `--protocol SSMTPA` (_port 465 with auth_)
- Mail intended to be **_delivered_ to DMS _from an external MTA_** (most tests): `--ehlo mail.external.tld` + `--from user@external.tld` + `--protocol ESMTP` (_uses port 25 implicitly, this is the default `--protocol` value_)
  - [x] `non-existing-user.txt`: Removable, only needs to send a generic mail with `--to nobody@example.test`.
  - [x] `amavis/spam.txt` + `rspamd/spam.txt` can be replaced by a generic mail that sends the standard GTUBE with `--from spam@external.tld`.
  - [ ] `amavis/virus.txt` + `rspamd/virus.txt` can likewise do the same with EICAR and `--from virus@external.tld`.
  - [ ] `emails/auth/` are all for testing the `SPOOF_PROTECTION=1` feature.
     - These are specifically testing the authenticated DMS account against the envelope sender (`--from`).
     - Sending a generic mail is sufficient, while `--auth-user` / `--from` can be provided separately as overrides.
  - [x] `emails/existing/` are all generic (_with the exception of the envelope-recipient via `--to`_) for testing Postfix features related to routing mail delivery.
    - `smtp_delivery.bats` tests against each mail by asserting against their `Subject` header.
    - [x] ~~Additionally `user-and-cc-local-alias.txt` needs to append the CC header `Cc: Existing Local Alias <alias2@localhost.localdomain>`~~
      - The alias (`alias2`) resolves to [`external1@otherdomain.tld`](https://github.com/docker-mailserver/docker-mailserver/blob/0889b0ff063a37b482113a684d934e4bd728a33c/test/config/postfix-virtual.cf#L4).
      - [ ] Alias resolution is asserted in test-case: `redirects mail to external aliases`, however that alias is used in another mail sent which makes the test-case **not reliable** for coverage on the CC header?
      - [x] **NOTE:** The file name incorrectly has a `local-alias` suffix while the alias is to an external domain? (_resolved since all files from this group were removed by a follow-up PR_)
- [ ] Docs permalink update: https://github.com/docker-mailserver/docker-mailserver/pull/3732#discussion_r1438737894
- [ ] Changing `external.tld` => `external-third-party.test` + `localhost.localdomain` => `example.test`, etc changes: https://github.com/docker-mailserver/docker-mailserver/pull/3732#discussion_r1438806746
- [ ] `postgrey.bats` whitelist test-case, investigate removing workaround with a separate client container and not sending mail direct to Postgrey port: https://github.com/docker-mailserver/docker-mailserver/pull/3732#issuecomment-1872678832
- [ ] `postscreen.bats` split container (needs resolvable sender address domain, and likely `TXT` DNS record for SPF after that): https://github.com/docker-mailserver/docker-mailserver/pull/3732#issuecomment-1873052202
- [ ] Instead of `-tlsc` (_or the long option name `--tls-on-connect`_), prefer setting `--protocol` with `SSMTP` or `SSMTPA` which is equivalent to specifying options `--auth PLAIN --port 465 -tlsc`. Although better alternative going forward would be to have a separate base config for swaks that includes the auth options. Usage in tests would then change to providing a different config with `--config path/to/swaks/config/with/auth`.

## Maintenance notes
- [ESMTP / EHLO and when it's important vs SMTP / HELO](https://github.com/docker-mailserver/docker-mailserver/pull/3732#discussion_r1438947970).
- Be mindful [when searching up docs for `swaks`](https://github.com/docker-mailserver/docker-mailserver/pull/3747#issuecomment-1878104440), the [correct docs source is here](https://github.com/jetmore/swaks/blob/develop/doc/base.pod) but appears to have had [broken Github rendering support since Jan 2017](https://github.com/jetmore/swaks/issues/90).
  - For `--data` option, the default template options are [covered here](https://github.com/jetmore/swaks/blob/073854d9ac87914487f1fd4ae6a87ce282375f73/doc/base.pod#data-options) (_**note:** permalink, this is rendered properly_).
  - When adding/modifying headers you have the `--add-header` and `--header` options, see the [`swaks` output + behaviour difference between these two here](https://github.com/docker-mailserver/docker-mailserver/pull/3747#issuecomment-1878104440).
    - `--header` will override an existing header if it exists, otherwise add the header, prefer it unless you need multiples of the same header.
    - Output shared is related to the `Message-ID` header and log snippets related to DMS which may be a useful reference.
      - Related was [another review comment](https://github.com/docker-mailserver/docker-mailserver/pull/3747#discussion_r1443892957) comparing Postfix Queue ID between rspamd and amavis/spamassassin along with rejection response from rspamd.
      - [Another review comment](https://github.com/docker-mailserver/docker-mailserver/pull/3747#discussion_r1446679671) details Postfix Queue ID generation. This is now included as a reference in the relevant test method that uses regex pattern to match the value.
- `postgrey_whitelist.txt` importance of blank line: https://github.com/docker-mailserver/docker-mailserver/pull/3732#discussion_r1438738526
- `postscreen.bats` context (_could probably improve the ENV docs as linked comment describes_): https://github.com/docker-mailserver/docker-mailserver/pull/3732#discussion_r1438753850
- The migration to swaks identified some mail headers in `DATA` that was discarded](_[`sieve/pipe`](https://github.com/docker-mailserver/docker-mailserver/pull/3732#discussion_r1438988130) and [`sieve/spam-folder`](https://github.com/docker-mailserver/docker-mailserver/pull/3732#discussion_r1438987853)_), these were mostly contextual / complimentary but could have been useful towards filtering. Presently [they are very simple](https://github.com/docker-mailserver/docker-mailserver/blob/b78978caedfa67e96425a57d4226cd1768474d17/test/tests/parallel/set3/mta/smtp_delivery.bats#L95-L96).
- `smtponly.bats`: [PR context](https://github.com/docker-mailserver/docker-mailserver/pull/3732#discussion_r1438970873) related to the [adjustment for swaks](https://github.com/docker-mailserver/docker-mailserver/blob/b78978caedfa67e96425a57d4226cd1768474d17/test/tests/parallel/set3/mta/smtponly.bats#L34-L43)
- `smtp_delivery.bats`: SMTPUTF8 test-case [PR context](https://github.com/docker-mailserver/docker-mailserver/pull/3732#discussion_r1438964728) for [change to leverage swaks features](https://github.com/docker-mailserver/docker-mailserver/blob/b78978caedfa67e96425a57d4226cd1768474d17/test/tests/parallel/set3/mta/smtp_delivery.bats#L266-L281).
- `tests.bats`: ENV `SPOOF_PROTECTION=1` auth test [PR context (forging test-case)](https://github.com/docker-mailserver/docker-mailserver/pull/3732#discussion_r1438955373) + [PR context (alias test-case)](https://github.com/docker-mailserver/docker-mailserver/pull/3732#discussion_r1438954584). [Current location](https://github.com/docker-mailserver/docker-mailserver/blob/b78978caedfa67e96425a57d4226cd1768474d17/test/tests/serial/tests.bats#L290-L321).

**Unrelated to swaks migration:**
- Tests for a regexp alias (`postfix-regexp.cf` support) coverage:
  - [Resolves to external address](https://github.com/docker-mailserver/docker-mailserver/pull/3732#discussion_r1438973042) could use a more appropriate address (_as linked review comment suggests_) to communicate the context better. [Current location](https://github.com/docker-mailserver/docker-mailserver/blob/b78978caedfa67e96425a57d4226cd1768474d17/test/tests/parallel/set3/mta/smtp_delivery.bats#L81).
  - [Resolves to local address](https://github.com/docker-mailserver/docker-mailserver/pull/3732#discussion_r1438972775), which differs by resolving to an internal address. [Current location](https://github.com/docker-mailserver/docker-mailserver/blob/b78978caedfa67e96425a57d4226cd1768474d17/test/tests/parallel/set3/mta/smtp_delivery.bats#L74-L75)
- Other aliases (`postfix-virtual.cf`) test coverage:
  - [Standard alias](https://github.com/docker-mailserver/docker-mailserver/pull/3732#discussion_r1438972227) (_[Current location](https://github.com/docker-mailserver/docker-mailserver/blob/b78978caedfa67e96425a57d4226cd1768474d17/test/tests/parallel/set3/mta/smtp_delivery.bats#L68-L69), incorrectly labeled as "external"_).
  - Standard alias [with recipient delimiter](https://github.com/docker-mailserver/docker-mailserver/pull/3732#discussion_r1438972443) (_[Current location](https://github.com/docker-mailserver/docker-mailserver/blob/b78978caedfa67e96425a57d4226cd1768474d17/test/tests/parallel/set3/mta/smtp_delivery.bats#L70-L71)_).
  - [Wildcard / Catch-All](https://github.com/docker-mailserver/docker-mailserver/pull/3732#discussion_r1438972766) 2nd domain aliased to a primary domain address ([Current location](https://github.com/docker-mailserver/docker-mailserver/blob/b78978caedfa67e96425a57d4226cd1768474d17/test/tests/parallel/set3/mta/smtp_delivery.bats#L72-L73)).
  - [This CC alias mail sent](https://github.com/docker-mailserver/docker-mailserver/blob/b78978caedfa67e96425a57d4226cd1768474d17/test/tests/parallel/set3/mta/smtp_delivery.bats#L91-L94) collides with a [similar alias test mail](https://github.com/docker-mailserver/docker-mailserver/blob/b78978caedfa67e96425a57d4226cd1768474d17/test/tests/parallel/set3/mta/smtp_delivery.bats#L82) where test-case would pass for either. There is no separate test coverage on the CC address. See reference section below for history details on why CC was used and what it's actually intended to test (_relevance is lost since switching from Dovecot LDA to LMTP_). [Related review comment](https://github.com/docker-mailserver/docker-mailserver/pull/3732#discussion_r1438973153).
- The [root mail test lacks contextual comment](https://github.com/docker-mailserver/docker-mailserver/pull/3732#discussion_r1438988499) ([Current location](https://github.com/docker-mailserver/docker-mailserver/blob/b78978caedfa67e96425a57d4226cd1768474d17/test/tests/parallel/set3/mta/smtp_delivery.bats#L97). Internal mail like update notifications or from Amavis / SA (via cronjob) needs to deliver mail without authentication and bypassing other checks, such as not requiring `PERMIT_DOCKER=container` and bypassing spam filtering. `sendmail` / `mail` commands will accomplish that and drop it straight into the Postfix delivery queue instead. `swaks` is not usable for that functionality AFAIK.
- [`privacy.bats` still relies on `--data` to send a generic mail](https://github.com/docker-mailserver/docker-mailserver/blob/b78978caedfa67e96425a57d4226cd1768474d17/test/tests/parallel/set3/mta/privacy.bats#L26-L32) with a [`User-Agent` header](https://github.com/docker-mailserver/docker-mailserver/blob/b78978caedfa67e96425a57d4226cd1768474d17/test/files/emails/privacy.txt#L3-L4), `swaks` should be able to remove that file and explicitly set the header in the test via `--header` option. [PR context](https://github.com/docker-mailserver/docker-mailserver/pull/3732#discussion_r1438989098).


## References

May still have some relevance/value when refactoring existing tests and with recent changes this context would potentially be even less obvious and not fun to track down 😅 

Original text files history:
- Grouped into `emails/existing/` folder in **PR 3732**: https://github.com/docker-mailserver/docker-mailserver/pull/3732
- Replaced with `swaks` equivalent commands in **PR 3747**: https://github.com/docker-mailserver/docker-mailserver/pull/3747

**`local-alias.txt`** vs **`external-alias.txt`** (_and similar variants where local/external is reversed in usage_):
- `postfix-accounts.cf` is [configured with a `user2@otherdomain.tld`](https://github.com/docker-mailserver/docker-mailserver/blob/0889b0ff063a37b482113a684d934e4bd728a33c/test/config/templates/postfix-accounts.cf#L2), which would include `otherdomain.tld` as a local DMS managed mail domain...thus locally managed (_at least it is these days_).
  - [Originally added in Oct 2015 by project author](https://github.com/docker-mailserver/docker-mailserver/pull/29).
  - [Follow-up PR](https://github.com/docker-mailserver/docker-mailserver/pull/30) shortly after introduces the first few "email templates", `amavis-virus` + `amavis-spam`, and the alias `external` + `local` usage but resolving to aliases to the opposite domain-part implied by the filenames.
  - [May 2016 introduces regexp aliases in tests](https://github.com/docker-mailserver/docker-mailserver/pull/194) with `local` and `external` variants correctly resolving to their implied domain-part via [`postfix-regexp.cf`](https://github.com/docker-mailserver/docker-mailserver/blob/0889b0ff063a37b482113a684d934e4bd728a33c/test/config/postfix-regexp.cf#L1C1-L1C1):
    - `local` variant: `test` + 2 digits prefix (_local-part prefix to a `localhost.localdomain` domain-part_) used in `regexp-alias-local.txt` (_as `test123@localhost.localdomain`_) has a [regexp alias to `user1@localhost.localdomain`](https://github.com/docker-mailserver/docker-mailserver/blob/0889b0ff063a37b482113a684d934e4bd728a33c/test/config/postfix-regexp.cf#L1C1-L1C1).
    - `external` variant: `bounce` (_local-part prefix to any domain-part_) used in `regexp-alias-external.txt` (_as `bounce-always@localhost.localdomain`_) has a [regexp alias to the same `external1@otherdomain.tld`](https://github.com/docker-mailserver/docker-mailserver/blob/0889b0ff063a37b482113a684d934e4bd728a33c/test/config/postfix-regexp.cf#L4).

**`user-and-cc-local-alias.txt`:**
- [Originally added July 2016](https://github.com/docker-mailserver/docker-mailserver/pull/240) to [resolve a multiple recipient delivery issue with Postfix + Dovecot](https://github.com/docker-mailserver/docker-mailserver/issues/230) (_alias usage is not relevant_). [Common enough issue online](https://wiki.archlinux.org/title/Dovecot#warning:_pipe_flag_%60D'_requires_dovecot_destination_recipient_limit_=_1).
- [Oct 2016 removed `dovecot_destination_recipient_limit=1` from Postfix `main.cf`](https://github.com/docker-mailserver/docker-mailserver/pull/363) due to ["unused parameter" warning spam in logs](https://github.com/docker-mailserver/docker-mailserver/issues/362).
  - The parameter name has a dynamic prefix and actually documented as [`transport_destination_recipient_limit`](https://www.postfix.org/postconf.5.html#transport_destination_recipient_limit), where the `transport` prefix maps to a transport name defined in `master.cf`. It is a parameter relevant to [`pipe(8)`](https://www.daemon-systems.org/man/pipe.8.html) (_Postfix delivery to external command_).
  - Earlier in Oct 2016, [Dovecot delivery was changed to LMTP instead of Dovecot LDA](https://github.com/docker-mailserver/docker-mailserver/pull/360) (_which was configured as the `dovecot` transport in `master.cf`_). Hence the warning log spam issue that followed.
  - As docs for the [Postfix `master.cf` `pipe` daemon](https://www.daemon-systems.org/man/pipe.8.html) describe under _"Single-Recipient Delivery"_, multi-recipient delivery can be problematic when `Delivered-to:` / `X-Original-To:` message headers are prepended.
    - Which `transport_destination_recipient_limit` can configure per-transport, whilst all other delivery agents (_`lmtp`, `pipe`, `smtp`, `virtual`_) otherwise get their default limit from [`default_destination_recipient_limit`](https://www.postfix.org/postconf.5.html#default_destination_recipient_limit) (_default: 50_).
    - Previously with the `dovecot` transport in `master.cf` this used [`flags=DRhu`](https://github.com/docker-mailserver/docker-mailserver/commit/a62062c382ef47abe49acf5805b8e8ccddcf0da0#diff-382e37a673d41779bffd7dfe054feb345c1f3e5f7121698f2a52e7e5322c7169L71) which as the `pipe` daemon docs (on flags section) states, requires the limit parameter is set to `1` to correctly work.
    - This configuration allowed supporting those specific headers with delivering to multiple recipients, with the only difference in [overhead by delivering to each recipient sequentially](https://narkive.com/VmgwEZZq.7) instead of in bulk to Dovecot to handle on it's end.
  - [Feb 2017 issue of the change to LMTP affecting Dovecot Sieve](https://github.com/docker-mailserver/docker-mailserver/issues/504), since the `X-Original-To:` header was not available, the mail that arrived for Sieve to process had already resolved the alias and lost this context that the additional header would have provided.
    - There are a few workarounds such as:
      - [Using another Postfix feature to conditionally match the recipient to prepend the header](https://narkive.com/P4OA8zgi.10), but it [removes privacy](https://dovecot.org/mailman3/archives/list/dovecot@dovecot.org/message/IQHRGOHSII6ZNCO442XEMF32DHRYIYAI/) and probably [doesn't work as desired with multiple recipients involved](https://github.com/42wim/til/blob/master/postfix/dovecot_and_envelope_recipient_address_for_sieve_filtering.md#first-failed-solution). Another comment explains some technical details [why delivering to multiple recipients had to be limited to processing one at a time](https://narkive.com/P4OA8zgi.4).
      - [Adjusting the sieve filter for different headers](https://github.com/docker-mailserver/docker-mailserver/issues/1861#issuecomment-804611454), but this is [not compatible with BCC recipients](https://github.com/docker-mailserver/docker-mailserver/issues/1861#issuecomment-1125998595). Another use-case is supporting a [catch-all address to direct mail to a folder via sieve script](https://serverfault.com/questions/624916/postfix-virtual-users-dovecot-sieve-envelop-does-not-contain-original-rec).
    - Besides `pipe`, it is [also supported by `local` and `virtual`](https://narkive.com/P4OA8zgi.2) delivery agents, but not `smtp` and `lmtp` delivery agents.
      - 7 years later with the [Postfix 3.3.0 release](https://www.postfix.org/announcements.html) (Feb 2018), the [`virtual`](https://www.postfix.org/virtual.8.html) and [`local`](https://www.postfix.org/local.8.html) transports gained support for [`enable_original_recipient`](https://www.postfix.org/postconf.5.html#enable_original_recipient) (default: yes), which prepends the `X-Original-To` header. This may have been for the ability to set `enable_original_recipient=no` according to the parameter docs.
      - In Mar 2020 the [Postfix 3.5 release](https://www.postfix.org/announcements.html) [`lmtp`](https://www.postfix.org/lmtp.8.html) and [`smtp`](https://www.postfix.org/smtp.8.html) delivery agents gained support for `Delivered-To:` / `X-Original-To:` header additions with the same `flags=DO` syntax that `pipe` supported, which likewise rely on `transport_destination_recipient_limit` being set to 1 to work.

**`catchall-local.txt`:**
- [Added July 2016](https://github.com/docker-mailserver/docker-mailserver/pull/241).
- Also [shows vhost for `otherdomain.tld`](https://github.com/docker-mailserver/docker-mailserver/pull/241/files#diff-c65d41b8ad46ee079a476545e7816994ee96c554eeac057e58c9f430f568b243R215), so that is treated as locally managed.

**`emails/root-email.txt`:**
- [`postmaster` local-part regexp alias with any domain-part to `user1@localhost.localdomain`](https://github.com/docker-mailserver/docker-mailserver/blob/0889b0ff063a37b482113a684d934e4bd728a33c/test/config/postfix-regexp.cf#L5)
- [Introduced April 2018](https://github.com/docker-mailserver/docker-mailserver/pull/952), testing the system `root` account delivers mail to the configured postmaster address (ENV `POSTMASTER_ADDRESS`). Test has mail delivered to root, normally aliased to `POSTMASTER_ADDRESS`, while the [regexp alias to user account](https://github.com/docker-mailserver/docker-mailserver/pull/952#issuecomment-382200172) is used as fallback.
- Neither `helper/setup.bash` or `smtp_delivery.bats` (_only user of `root-email.txt`_) sets the `POSTMASTER_ADDRESS` ENV. This will potentially be set in future which may affect this in a subtle way? :shrug:

`existing/user1` was a common file used for generic mail testing. With the switch to `swaks` this is roughly the defaults, but there was still some inconsistencies (_sometimes intentional by specifying defaults explicitly as options for visibility_) vs how it used to be.

---

## Swaks related reference

Most mail can use the following as base config and override via CLI options as needed (_this was initial work towards advising a swaks config file for base/defaults_):

```
# Envelope sender and recipient:
--from: user@external.tld
--to: user1@localhost.localdomain
# Sender and recipient mail headers:
--h-From: "External mail client" <user@external.tld>
--h-To: "Existing Local User" <user1@localhost.localdomain>
```

Prior to `swaks` we [relied on netcat / nc to stream a series of SMTP commands with the common format](https://github.com/docker-mailserver/docker-mailserver/pull/3732#issuecomment-1872667197):

```
HELO mail.external.tld
MAIL FROM: user@external.tld
RCPT TO: <recipient here>
DATA
<data content here>

.
QUIT
```

We switched to ESMTP with EHLO for all mail when adopting swaks, previously that was only used when necessary (eg: SMTP Auth).
